### PR TITLE
[FIX] core: fix cached translations

### PR DESCRIPTION
**Pull request by HydrionBurst**
before this commit:
translations updated by `update_field_translation` api cannot be detected by t-cache and some cached data whose model overrides `write` with an extra 'clear_caches()'

after this commit:
`update_field_translation` will call `write` is the translation can be updated it does the following 3 important things
1. security check
2. update write_date if needed to support t-cache
3. clear_cache by the override of `write`

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by rdeodoo — APPROVED** [review]
Approving for the website case, since the test was taken from https://github.com/odoo/odoo/pull/124192/commits/028089b0f09d47ff0bc5c9f95018c62e77a2bf16 and the branch is now green, our use case is solved.

Don't forget to add the `opw-3305117` in your commit message.

A `Step to reproduce` is never too much, you might want to include the one I already wrote in my commit message, even if website oriented, but I'll let that up to you.

File: addons/website/tests/test_menu.py [hunk 1]
```diff
@@ -185,3 +189,35 @@ def test_02_menu_anchors(self):
...
+            'website_id': website.id,
+        })
+        self.env['ir.module.module']._load_module_terms(['website'], [fr.code])
+
+        # Load cache
+        self.url_open('/%s' % fr.url_code)
+        self.url_open('/%s?edit_translations=1' % fr.url_code)
+
+        # Translate
+        root = html.fromstring(menu.mega_menu_content, parser=html.HTMLParser(encoding="utf-8"))
```
Comment (HydrionBurst) [inline addons/website/tests/test_menu.py:212]: Hi, @rdeodoo 
could you help me check if the `parser=html.HTMLParser(encoding="utf-8")` is still necessary here?

Comment (rdeodoo) [reply addons/website/tests/test_menu.py:212]: It's useless.
I copy pasted this line when I wrote this part of the test (with the `sha256` below).
The default `parser` used in `lxml` is already the `HTMLParser` if omitted.

Also, about the encoding, it's probably useless, from what I know in python3, utf-8 is generally the default in most libs.

**Review by rco-odoo — CHANGES_REQUESTED** [review]
The proposed patch looks like it contains some refactoring of code, and I am not even sure all this is necessary for the fix.  And the refactoring does not look correct to me.

File: odoo/models.py [hunk 2]
```diff
@@ -3086,7 +3082,7 @@ def _update_field_translations(self, field_name, translations, digest=None):
                     }
                 new_translations[lang] = field.translate(translation.get, old_value)
             self.env.cache.update_raw(self, field, [new_translations], dirty=True)
-            self.modified([field_name])
+        self[field_name] = self[field_name]
```
[the remark this answers is not in the document]
Comment (HydrionBurst) [reply odoo/models.py:3085]: > This side effect might disappear some day

do you mean in some day, when we call `record.name =  new_name` and if the `new_name` is exactly the same as the `old_name` in cache, the `name` field won't be marked modified?

**Comment by HydrionBurst** [comment]
@rco-odoo I have added a test which will fail if `modified` is not triggered
Could you review it again?

**Comment by HydrionBurst** [comment]
@rco-odoo Code is refactored, could you review it again?

**Comment by rco-odoo** [comment]
@robodoo override=ci/security

The added import of `hashlib` is only used for tests.

**Pull request closed without merging**